### PR TITLE
Async Demotion (Part 1)

### PR DESCRIPTION
**Pull request by kavon**
For chains of async functions in the call graph where suspensions can be statically
proven to never be required, this pass removes all suspensions and
turns the functions into synchronous functions.

For example, this function does not actually require any suspensions,
once the correct executor is acquired upon initial entry:

```swift
func fib(_ n: Int) async -> Int {
  if n <= 1 { return n }
  return await fib(n-1) + fib(n-2)
}
```

So we can specialize the above into this for better performance:

```swift
func fib(_ n: Int) async -> Int {
  return fib_sync(n)
}

func fib_sync(_ n: Int) -> Int {
  if n <= 1 { return n }
  return fib(n-1) + fib(n-2)
}
```

We then rewrite callers of `fib` to use the `sync` entry-point
when we can prove that the caller  will invoke `fib_sync` on a matching executor. As a result, callers of `fib_sync` can also remove the hop-back after the call.

By rewriting callers to use the sync entry, that reduces the number of required suspensions, which can lead to more recursive specializations, etc.

TODO: more details.

### CODE REVIEW

File: SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AsyncDemotion.swift [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+  var knownAsyncApplys: [ApplySite] = []
+  var unknownAsyncOp: Instruction? = nil
+
+  for inst in function.instructions {
+    // Track hops in a low-resolution way.
+    //
+    // FIXME: would be better to associate each call-site with the executor
+    // it would happen under, so demoted entry-points can be called from
+    // multi-executor async functions.
+    if let hop = inst as? HopToExecutorInst {
```
Comment (eeckstein) [inline SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AsyncDemotion.swift:236]: you can formulate this as a switch:
```
switch inst {
case let hop as HopToExecutorInst:
   ...
case is GetAsyncContinuationInst,
     is GetAsyncContinuationAddrInst,
     is AwaitAsyncContinuationInst:
...

File: SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AsyncDemotion.swift [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+        fatalError("could not find def for value:\n \(val)")
+      }
+      return result
+    }
+
+    internal mutating func rootDef(value: Value,
+                                   path: SmallProjectionPath) -> WalkResult {
+      log("FindExecutor.rootDef -> \(value)")
+
+      if let arg = value as? FunctionArgument {
```
Comment (eeckstein) [inline SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AsyncDemotion.swift:55]: You can use a switch here (see my other comment)

**Comment by kavon** [comment]
@swift-ci please smoke test

**Comment by kavon** [comment]
@swift-ci please smoke test

**Comment by kavon** [comment]
@swift-ci Please smoke test compiler performance

File: SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AsyncDemotion.swift (hunk 2 above)
Comment (kavon) [reply SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AsyncDemotion.swift:118]: this conditional block got more complex than a switch can handle I think

**Comment by kavon** [comment]
@swift-ci please smoke test and merge

**Comment by kavon** [comment]
@swift-ci please smoke test and merge

**Pull request merged**